### DISCUSSION TITLE
Quantum random number generator

### DISCUSSION THREAD

**Discussion by rdagger** [ozel/DIY_particle_detector#35] [Q&A]
I built your particle detector (BPX61 version with glass removed).  Thanks for taking the time to publish this great project!  My goal is to build a quantum random number generator based on radioactive decay as a source of entropy.  I’m trying to test the detector using a 25 mm diameter glass marble doped with 3% Uranium 238 that I purchased from United Nuclear.  I’m not getting any pulses on my hardware oscilloscope and I don’t hear anything but noise when piped to my computer’s sound card.  I followed your troubleshooting guide and I believe the hardware is assembled correctly.

I’m assuming that there is a problem with my testing methodology or that the uranium concentration in the marble is too low to register on the detector.  I’d really appreciate any suggestions.

Here are some pics:
![Case01r](https://user-images.githubusercontent.com/106355/113444410-b4150f80-93a8-11eb-9fd2-3b5e184557ed.JPG)
![Case02r](https://user-images.githubusercontent.com/106355/113444470-d149de00-93a8-11eb-9beb-845ff0ec81a2.JPG)
![Case03r](https://user-images.githubusercontent.com/106355/113444491-d870ec00-93a8-11eb-9e1a-c7e70f729723.JPG)

**Comment by ozel** [comment]
That is a great build! it looks totally awesome. Easily wins the MacGyver award, hands down! :D 
May I resued the pictures? If so, please let me know how to cite them.

From uranium glass, there are not many alphas if at all, usually. But you should still be able to detect quite some electrons and gamma photons with the BPX61 version. The electron detector variant with more diodes would give you 4 times more events.

Which troubleshooting steps do not work for you? Does the detector react to light when the case is opened?
Also check if the BPX61metal case is maybe touching the circuit board. There is a chance for a short circuit if it rests evenly on the board surface. If so, melt the solder again and lift it up just for a millimeter. 

BTW, later for analysing the randomness of you detector data, you can checkout this nice website: https://www.fourmilab.ch/hotbits/

**Reply by rdagger** [reply]
Please feel free to use the pictures.  You can cite them to rdagger.  Once I get it working I will post a video on my YouTube channel:  https://www.youtube.com/rdagger68 and on my website www.rototron.info.

I didn't have any issues troubleshooting that's why I think it is the marble.  Although I am seeing a lot of noise.  I'll post a few oscilloscope pictures in a few days.  There is 10K ohms across the battery leads.  I see a difference on the scope when light is present and there is a ping on the scope when I tap the case.  I used a 3D printed spacer when I soldered the BPX61 to ensure its case was isolated from the PCB.  

Thanks for the link and the feedback!

**Comment by rdagger** [comment]
I installed a new 9.6V battery and replaced my scope leads with a 30 cm BNC cable plugged directly into the detector.  This fixed the noise problem.  I'm now getting readings on my scope that look like your samples.

![Alpha Detection 2021-04-13](https://user-images.githubusercontent.com/106355/114619568-9032b300-9c5f-11eb-8179-64f509d2a24b.PNG)

Unfortunately, my doped marble is only emitting about 3 pulses per minute so I need to find a better source.  I know your site dissuades  the use of smoke detector 241AM disks but I've done some research and it appears relatively safe as long as you don't try to remove the isotope from the disks.  The americium is electroplated onto the metal disks and is protected by a thin, gold-metal layer that is very chemically resistant.  There have been 2 documented cases of people swallowing the disks without health issues despite over 2 weeks in the GI tract.

Here are my references:
_Intake and Subsequent Fate of a Ceramic Particle Containing 2.85 μCi 241Am
Smith, L. R.; Sullivan, P. A.; Laferriere, J.; Cumming, E.; Demis, D.
Health Physics: April 1983 - Volume 44 - Issue 4 - p 329-334_

_Ingestion of 241Am Sources Intended for Domestic Smoke Detectors
Rundo, J.; Fairman, W. D.; Essling, M.; Huff, D. R.
Health Physics: December 1977 - Volume 33 - Issue 6 - p 561-566_

**Reply by SteveF161** [reply]
Get some fiestaware pottery shards. I got some for about GBP 20 (...a rip-off, but unavoidable. ) See my blog at [https://clanhouse.com/cern-at-home-ii-more-diy-radiation-detectors](https://clanhouse.com/cern-at-home-ii-more-diy-radiation-detectors )

**Reply by rdagger** [reply]
I'd prefer an intact piece to shards.  Maybe I can find a small teapot cover that fits in my case.  Thanks for your response.

btw: I reviewed the [U.S.NRC regulations](https://www.nrc.gov/materials.html) and it looks like it is illegal to possess americium outside of a smoke detector or an ionizing radiation measuring instrument containing an internal calibration source.
(based on [Part 30 - Rules of General Applicability to Domestic Licensing of Byproduct Material](https://www.nrc.gov/reading-rm/doc-collections/cfr/part030/index.html))

**Reply by ozel** [reply]
great job, the pulse looks exactly as it should! Shielded cables are mandatory, yes.

you found already the reason why I can't really recommend opening old smoke detectors[1]. but I fully agree with your health risk assessment, the implanted metal foil is actually safer than chips of U238 glaze. There are small Uranium glazed objects like the pendant from the Majolika manufacture shown in the readme, but it can be a bit annoying to hunt for them. 
 
A legal way is purchasing one of those check sources with Po210 or Am214 (better because of longer half life), they are sometimes on eBay too and usually below national exemption limits (e.g. https://www.spectrumtechniques.com/products/sources)

Just keep in mind that you will observe a pile-up of pulses once you reach a few kilo Bequeerel of activity close to the diode.  But if you don't depend on using a low-cost soundcard you can easily shrink the pulse width by modifying the circuit. 
The feedback resistor of the first opamp can be reduced a lot as long as you still get enough signal-to-noise ratio on your scope/data acquisition interface. Basically, you can use the same resistor and capacitor values as in the electron detector version. This makes the pulses smaller by a factor of about 10. For your randomness generator, the pulse amplitude (=energy) won't matter. Only the timing between pulses should be measured accurately. In that respect, I am still not sure what advantage you gain from only focusing on alphas instead of going with the electron detector variant using 3 more diodes. 
Because the diodes are so small, background pulse rates from terrestrial and cosmic radiation are low in general. Even the alpha spectrometer setup will see an occasional radon decay alpha and even within your box. On the other hand, any real radiation background should be uncorrelated with your source and therefore not impact the statistics. 
Also long as electronic noise from a poor signal-to-noise ratio doesn't impact your statistics data, I'd think the particle source should not matter. I might be missing something though.

If you want to do a quick alpha test, do the 'radioactive ballon' experiment in your basement. Search my other replies for that, it's pretty hilarious. I really need to make a wiki page on sources soon! 
With some more tinkering, you could build a continuous radon sampler as a randomness source! Haven't seen anything like that so far. It should be a bit correlated with the local weather and outgassing conditions of the soil, but that's actually also quite exciting if you ask me. :-)

[1]: On Aliexpress you can still find 'ion chambers' for 1$, containing the discs you described...

**Reply by rdagger** [reply]
I’m sure you’re right that the electron detector is a better fit for QRNG.  I’ll order the BPW34F photodiodes.  Since I’m ordering parts, is there a chip you can recommend to integrate the detector with a microcontroller such as an ESP32 or Raspberry Pi Pico?  I’m thinking maybe a precision ADC that is optimized for the output range of the detector.

**Reply by ozel** [reply]
late reply, but integrated ADCs are often too non-linear and noisy for this application, the ESP ones got particularly poor performance, the RPi Pico one is similar and more geared towards general purpose use I believe. 
I had some success with the SAMD series used in certain Arduino boards because some of those microcontrollers allow lowering the internal reference voltage for the ADC which increases the usable range.

I've had very good success with the ADS1115 ADCs available on breakout boards, not only because of the high 16-bit resolution, rather due to the internal gain amplifier. 
In any case, you have to shield the signal cable really (e.g. microphone/coax cable) well until the ADC input pin or put everything, with a separate battery for the microcontroller, into the same metal box. That's the easiest approach to get it working. 
Once pulse recoding works you can start backtracking and see what kind of power supply filtering/cable shielding is required for powering externally (hint: it can be a nightmare ;).

**Reply by rdagger** [reply]
Thanks!  I did order an ADS1220 but if that doesn't work I'll try the 1115.

**Reply by rdagger** [reply]
I found a vintage Kodak 814 camera on eBay for $10 US.  I believe the Ektar lens is doped with thorium dioxide.  It works with the alpha detector but I switched to the electron detector because I'm getting better results with the ADC.  

![Case with Lens (2s)](https://user-images.githubusercontent.com/106355/131379743-56cd72ce-4164-4231-83db-10cc26b4001f.jpg)

![Thorium Lens Electron Detector Scope b](https://user-images.githubusercontent.com/106355/131380201-416e8d10-d188-4d8d-b7e8-f7a5a45c2b00.jpg)

I still need to work on the ADC.  I'm using an ADS1220 but it only has a delta-sigma sampling rate of 2000 SPS (samples per second) which is slow for 100 μs pulses.  You mentioned you had good success with the ADS1115 which is 860 SPS.  Do you have any strategies for capturing the pulses with an ADC?

**Reply by ozel** [reply]
nice! if you see browning or a yellow taint while looking through the lens it is very likely thoriated (so-called color centers, due to the radiation). Sometimes those lenses emit even alphas right from the surface of the glass.

For the alpha detector, ~50k SPS are needed for recording the whole pulse (deducing energy spectra).
For the electron detector, it is enough to just trigger on pulses and count them, so much lower rates are fine. In that case, internal ADCs come in handy because they are quite fast and linearity/accuracy doesn't matter for detecting pulse edges.
SAMD21 and SAMD51 on Adafruit/Arduino zero/M1/M4 boards work quite well for counting fast pulses from the electron detector.
Keep in mind that the sampling rate setting usually messes with the ADC input impedance which in turn affects the pulse size seen by the ADC: https://blog.thea.codes/getting-the-most-out-of-the-samd21-adc/.
The oscilloscope input impedance is usually fixed at 1 MOhm but ADCs may have 10 to 100 times less, "eating" into the signal. So at the ADC, pulse amplitudes shrink but the negative edge slope is becoming less steep (=slower), too. I thought the ADS1115 was fast enough to catch the negative electron detector pulse edge reliably (not the whole pulse shape) but my memory might be wrong. It also has a comparator mode, that can generate an interrupt. Maybe I was using that.

**Reply by rdagger** [reply]
Thanks for the comparator tip, which is very helpful!  The ADS1220 doesn’t have a comparator so I’ll try out the ADS1115.  I’m also experimenting with the SGTL5000 which is a low power stereo codec chip that has analog input, amplified output and an I2S port.

**Reply by ozel** [reply]
The SGTL5000 is certainly a great choice for recording whole alpha pulses.

But again, please don't waste time trying to record complete pulse shapes if you use the electron detector. While the pulse size is still proportional to the _deposited_ electron energy, it is usually incomplete because some electrons exit the diode after depositing only a fraction of their energy. You won't get a useful energy spectrum or anything interesting out of the electron pulse shapes (at least when recording from typical gamma check sources, or very mixed particle sources like uranium glass or a thoriated lens). See the discussion about figure 7b in https://www.mdpi.com/1424-8220/19/19/4264.

The ADS1115 comparator is still relying on the ADC, btw., meaning it will be just a bit faster if you use the external interrupt function compared to evaluating the ADC values for certain thresholds in software.

I would be curious to know how exactly it did not work with the ADS1220. My approach would be to connect an oscilloscope in parallel and print out short ADC value arrays via Arduino serial plotter once a certain threshold got underrun.

**Reply by rdagger** [reply]
I did get the ADS1220 working but I wasn’t 100% confident that I was catching all the pulses.  This project has been on the back burner for several weeks, but it felt like the audio count averaged higher than the ADC count.  It could be because I wrote the [ADS1220 driver](https://github.com/rdagger/micropython-ads1220) and I’m using MicroPython.  That’s the reason I like the idea of the comparator to compensate for the interpreted code.  The reason I like the SGTL5000 is because it can provide audio feedback without any programming overhead.

Here's a Dungeons & Dragons dice demo using the quantum random number generator powered by an ESP32 running MicroPython.
![D D Dice Sample01](https://user-images.githubusercontent.com/106355/139540979-deb6699e-ac8d-43bb-8558-91bc07edab31.gif)
:
